server/aesdsocket.c

```diff
@@ -0,0 +1,314 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <stdbool.h>
+
+#include <fcntl.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+
+#include <sys/stat.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <netdb.h>
+#include <syslog.h>
+#include <signal.h>
+
+#define PORT                "9000"
+#define BACKLOGS            5
+#define OUTPUT_FILE_PATH    "/var/tmp/aesdsocketdata"
+#define MAX_BUFFER_SIZE     256
+
+
+int socket_fd;
+int signal_caught = 0;
+char host[NI_MAXHOST];     // for getting connection logs
+
+static void signal_handler(int sig) {
+    signal_caught = sig;
+    shutdown(socket_fd, SHUT_RDWR); // shutdown socket
+}
+
+
+static int send_data_to_client(int socket, FILE * fd) {
+    
+    int ret = 0;
+    int  bytes_read;
+    char *packet_ptr;
+    size_t sent_cnt;
+    char read_buffer[MAX_BUFFER_SIZE];
+    memset(read_buffer, 0, MAX_BUFFER_SIZE);
+    while (!feof(fd))
+    {
+        bytes_read = fread(read_buffer, 1, MAX_BUFFER_SIZE, fd);
+        packet_ptr = read_buffer;
+        sent_cnt = send(socket, packet_ptr, bytes_read, 0);
+        if (sent_cnt == -1) {
+            syslog(LOG_ERR, "Unable to send data to client\n");
+            ret = -1;
+            break;
+        }
+        memset(read_buffer, 0, MAX_BUFFER_SIZE);
+        syslog(LOG_INFO, "Sent %lu bytes to client\n", sent_cnt);
+    }
+
+    return ret;
+}
+
+static int read_packet(int socket, FILE * fd) {
+    
+    int result = 0;
+    char read_buffer[MAX_BUFFER_SIZE];
+    memset(read_buffer, 0, MAX_BUFFER_SIZE);
+    int continue_read = 1;
+    
+    while(continue_read) {
+        //read from socket
+        ssize_t byte_read = recv(socket, read_buffer, (MAX_BUFFER_SIZE-1), 0);
+        if(byte_read == -1) {
+            continue_read = 0;
+            result = -1;
+            syslog(LOG_ERR, "Error reading data from socket\n");
+        } else  if(byte_read == 0) {  
+            continue_read = 0;
+            result = -1;
+        } else {
+            syslog(LOG_ERR, "Received %lu bytes from client\n", byte_read);
+            // check for \n
+            char *nl_char = memchr(read_buffer, '\n', MAX_BUFFER_SIZE);
+            continue_read = (nl_char == NULL) ? 1:0;
+
+            // write data to file /var/tmp/aesdsocketdata
+            fwrite(read_buffer, byte_read, 1, fd);
```

Comment:
@KapureCUB with this design you will write to the file before the packet completes, which may cause problems with future assignments.